tls/s2n_x509_validator.c

```diff
@@ -371,6 +371,39 @@ s2n_cert_validation_code s2n_x509_validator_validate_cert_chain(struct s2n_x509_
 
         S2N_ERROR_IF(op_code <= 0, S2N_ERR_CERT_UNTRUSTED);
         validator->state = VALIDATED;
+
+        /* Check to ensure certificate signatures appear in certificate_signature_preferences.
+         * Additionally we only want this check to apply in TLS1.3 */
+        if (conn->config->security_policy->certificate_signature_preferences &&
+                conn->actual_protocol_version >= S2N_TLS13) {
+            
+            STACK_OF(X509) *validated_chain = X509_STORE_CTX_get1_chain(validator->store_ctx);
+
+            if (!validated_chain) {
+                goto clean_up;
```

Comment:
Can you use DEFER_CLEANUP instead, possibly with a wrapper around the existing wipe_cert_chain method if necessary? That'll make your error handling easier.